### PR TITLE
Expose feed provider options

### PR DESCRIPTION
**Pull request by SuperDragonXD**
## Description

This PR exposes the (previously hidden) Feed Provider settings. It also re-adds the "ignore feed whitelist" debug flag.

Most of the code (except compose UI) is ported from the previous V2 implementation of custom feeds.

should hopefully build

## Type of change

:x: General change (non-breaking change that doesn't fit the below categories like copyediting)
:x: Bug fix (non-breaking change which fixes an issue)
:white_check_mark: New feature (non-breaking change which adds functionality)
:x: Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: lawnchair/src/app/lawnchair/preferences/PreferenceManager.kt [hunk 1]
```diff
@@ -51,7 +51,8 @@ class PreferenceManager private constructor(private val context: Context) : Base
 
     val drawerOpacity = FloatPref("pref_drawerOpacity", 1F, recreate)
     val coloredBackgroundLightness = FloatPref("pref_coloredBackgroundLightness", 0.9F, recreate)
-    val feedProvider = StringPref("pref_feedProvider", "")
+    val feedProvider = StringPref("pref_feedProvider", "", recreate)
```
Comment (MrSluffy) [inline lawnchair/src/app/lawnchair/preferences/PreferenceManager.kt]: Maybe we can add restart here? or take a warning when switching to feed may need lawnchair to restart?

**Comment by MrSluffy** [comment]
And got this error

to reproduce, 
1. apply any feed provider ex SmartSpacer once applied, 
2. unintall smartspacer
3. and open home settings error occured

```
Lawnchair (Debug) bug report Dec 19, 2023 7:01:28 AM
version: 14 Dev (54fd1a9) (14)
commit: 54fd1a9
build.brand: Xiaomi
build.device: sweet
build.display: UP1A.231105.003
build.fingerprint: Redmi/sweet_eea/sweet:13/TKQ1.221013.002/V14.0.7.0.TKFEUXM:user/release-keys
build.hardware: qcom
build.id: UP1A.231105.003
build.manufacturer: Xiaomi
build.model: M2101K6G
build.security.level: 2023-11-01
build.product: aosp_sweet
build.type: user
version.codename: REL
version.incremental: 1701149661
version.release: 14
version.sdk_int: 34
display.density_dpi: 440
isRecentsEnabled: true

error: Uncaught exception

java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at jdk.internal.util.Preconditions.checkIndex(Preconditions.java:266)
	at java.util.Objects.checkIndex(Objects.java:359)
	at java.util.ArrayList.get(ArrayList.java:434)
	at app.lawnchair.ui.preferences.components.FeedPreferenceKt.FeedPreference(FeedPreference.kt:71)
	at app.lawnchair.ui.preferences.destinations.HomeScreenPreferencesKt$HomeScreenPreferences$1$1.invoke(HomeScreenPreferences.kt:89)
	at app.lawnchair.ui.preferences.destinations.HomeScreenPreferencesKt$HomeScreenPreferences$1$1.invoke(HomeScreenPreferences.kt:68)
```

File: lawnchair/res/values/strings.xml [hunk 2]
```diff
@@ -190,6 +190,9 @@
       <!-- <string name="gesture_double_tap" /> -->
       <string name="minus_one_enable">Feed</string>
       <string name="minus_one_unavailable">No feed installed.</string>
+      <string name="feed_provider">Feed Provider</string>
+      <string name="feed_default">Default</string>
+
```
Comment (Goooler) [inline lawnchair/res/values/strings.xml:195]:
```suggestion
```

File: lawnchair/src/app/lawnchair/ui/preferences/components/FeedPreference.kt [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                icon = it.loadIcon(context.packageManager),
+            )
+        }
+
+    return entries
+}
+
+fun getProvidersList(context: Context): List<ListPreferenceEntry<String>> {
+    val entries = getEntries(context)
+    return entries.map {
```
Comment (Goooler) [inline lawnchair/src/app/lawnchair/ui/preferences/components/FeedPreference.kt]:
```suggestion
    getEntries(context) entries.map {
```

File: lawnchair/src/app/lawnchair/ui/preferences/destinations/HomeScreenPreferences.kt [hunk 4]
```diff
@@ -83,6 +85,9 @@ fun HomeScreenPreferences() {
                 description = if (feedAvailable) null else stringResource(id = R.string.minus_one_unavailable),
                 enabled = feedAvailable,
             )
+            if (feedAvailable && prefs2.enableFeed.getAdapter().state.value) {
+                FeedPreference(context = context)
+            }
```
Comment (suphon-t) [inline lawnchair/src/app/lawnchair/ui/preferences/destinations/HomeScreenPreferences.kt]: Use `ExpandAndShrink` here

**Comment by suphon-t** [comment]
The default option shouldn't be here. Just Google and Smartspacer.

Screenshot_20231219-120441

File: lawnchair/src/app/lawnchair/ui/preferences/destinations/HomeScreenPreferences.kt [hunk 5]
```diff
@@ -61,6 +62,7 @@ fun HomeScreenPreferences() {
     val prefs = preferenceManager()
     val prefs2 = preferenceManager2()
     val scope = rememberCoroutineScope()
+    val context = LocalContext.current
```
Comment (suphon-t) [inline lawnchair/src/app/lawnchair/ui/preferences/destinations/HomeScreenPreferences.kt]: Move the context into `FeedPreference`

File: lawnchair/src/app/lawnchair/preferences/PreferenceManager.kt (hunk 1 above)
Comment (suphon-t) [reply lawnchair/src/app/lawnchair/preferences/PreferenceManager.kt:54]: I think we should make it apply without restarting instead. That would be a better user experience.

Comment (suphon-t) [reply lawnchair/src/app/lawnchair/preferences/PreferenceManager.kt]: I think we should make it apply without restarting instead. That would be a better user experience.

**Review by MrSluffy — APPROVED** [review]

**Review by Goooler — APPROVED** [review]

**Comment by suphon-t** [comment]
> The default option shouldn't be here. Just Google and Smartspacer.
> 
> Screenshot_20231219-120441

@SuperDragonXD

**Comment by SuperDragonXD** [comment]
Regarding this, it is the same behavior as Lawnchair 10. We could probably remove the "Google" option in the whitelist and replace it with the default one (could be renamed to Google Feed).

**Comment by suphon-t** [comment]
> Regarding this, it is the same behavior as Lawnchair 10. We could probably remove the "Google" option in the whitelist and replace it with the default one (could be renamed to Google Feed).
> 
> 

Yeah, I think that works. But what should happen if there are no feed providers installed?

**Comment by SuperDragonXD** [comment]
It should just disable the "Feed" option (with the description: "No feed provider installed) and hide the "Feed Provider" preference.

**Review by suphon-t — APPROVED** [review]

**Pull request merged**